### PR TITLE
Abstract sockets

### PR DESCRIPTION
**Pull request by Mixaster995**
## Issue
part of https://github.com/networkservicemesh/sdk-vpp/issues/370
part of https://github.com/networkservicemesh/integration-k8s-kind/issues/325

### CODE REVIEW

**Comment by edwarnicke** [comment]
Have you tested this with cmd-forwarder-vpp ?  My initial testing with it is failing (though need to check to see if I'm adapting to it correctly to be sure).

**Comment by Mixaster995** [comment]
@edwarnicke yes, i tested it on equinix metal, everything worked fine. For correct testing we need to update and build images for  `cmd-forwarder-vpp`, `cmd-nsc-vpp`, `cmd-icmp-responder-vpp`

**Comment by edwarnicke** [comment]
@Mixaster995 Good to know its working on equinix metal.  My point is: could you push your code that fixes up cmd-forwarder-vpp for this change so I can try running its unit tests to make sure they are working as expected?

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -34,6 +34,10 @@ jobs:
       - name: Build
         run: go build -race  ./...
       - name: Test
+        if: matrix.os != 'windows-latest'
+        run: sudo -E PATH="$PATH" bash -c "go test -race ./..."
+      - name: Test
+        if: matrix.os == 'windows-latest'
```
Comment (edwarnicke) [inline .github/workflows/ci.yaml:40]: Could you explain why this change?

File: pkg/networkservice/mechanisms/memif/common.go [hunk 2]
```diff
@@ -33,60 +36,96 @@ import (
...
+	api.Connection
+}
+
+var (
+	netNS     netns.NsHandle
+	netNSPath string
+)
+
+// nolint:gochecknoinits
+func init() {
```
Comment (edwarnicke) [inline pkg/networkservice/mechanisms/memif/common.go:58]: Why are we doing this in an init() ?

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (Mixaster995) [reply .github/workflows/ci.yaml:40]: this test https://github.com/Mixaster995/sdk-vpp/blob/abstract-sockets/pkg/tools/proxy/proxy_test.go#L44
failes without these lines

File: pkg/networkservice/mechanisms/memif/common.go (hunk 2 above)
Comment (Mixaster995) [reply pkg/networkservice/mechanisms/memif/common.go]: made refactoring, removed `init` method

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (edwarnicke) [reply .github/workflows/ci.yaml:40]: Got it :)

File: pkg/networkservice/mechanisms/memif/common.go [hunk 3]
```diff
@@ -33,60 +37,96 @@ import (
...
+	isExternal bool
+
+	api.Connection
+}
+
+var (
+	netNS     netns.NsHandle
+	netNSPath string
+	once      sync.Once
+)
```
Comment (edwarnicke) [inline pkg/networkservice/mechanisms/memif/common.go:57]: @Mixaster995 these should not be global variables, they should be attributes of the chain(s) that use them.

Comment (Mixaster995) [reply pkg/networkservice/mechanisms/memif/common.go]: @edwarnicke made refactoring to remove usage of global variables - please, take a look

File: pkg/networkservice/chains/forwarder/server.go [hunk 4]
```diff
@@ -79,6 +79,7 @@ func NewServer(ctx context.Context, name string, authzServer networkservice.Netw
 	)
 	nsClient := registryclient.NewNetworkServiceRegistryClient(ctx, clientURL, registryclient.WithDialOptions(clientDialOptions...))
 
+	netNsInfo := memif.NewNetNSInfo()
```
Comment (edwarnicke) [inline pkg/networkservice/chains/forwarder/server.go:82]: @Mixaster995  Why are we doing this outside the memif.NewServer() ?  If this can be safely done inside memif.NewServer it vastly simplifies the use of memif.NewServer...

Comment (Mixaster995) [reply pkg/networkservice/chains/forwarder/server.go]: @edwarnicke i was under impression that it was required for some logic, but was mistaken. Made some refactoring and tested all changes.

Comment (edwarnicke) [reply pkg/networkservice/chains/forwarder/server.go]: @Mixaster995 Its all good.  When reworking someone else's work, its sometimes hard to figure out if something that looks weird is essential or not.

**Review by edwarnicke — APPROVED** [review]

**Review by denis-tingaikin — APPROVED** [review]
